### ISSUE TITLE
Add support for running Unit Tests in Gradle project with IntelliJ builder/runner  

### ISSUE THREAD

**Issue by asodja** [quarkusio/quarkus#8737]
**Description**
Currently it seems that if you want to run Gradle project with IntelliJ runner this does not work. So if I setup Gradle project with `Build and run using IntelliJ` and `Run tests using IntelliJ` (reference https://www.jetbrains.com/help/idea/work-with-tests-in-gradle.html#configure_gradle_test_runner) I get an exception like:
```
java.lang.RuntimeException: java.lang.RuntimeException: The test class class org.my.group.MyResourceTest is not located in any of the directories [classes/scala/native-test, classes/java/test, classes/scala/test, classes/java/native-test, /test-classes, bin/test, classes/kotlin/test, classes/kotlin/native-test)
```
This was tested with example Gradle project (https://quarkus.io/guides/gradle-tooling#project-creation). 

**Feature request**

It would be great that Gradle runner would not be needed in IntelliJ. Running Unit test in a sample Gradle project with Gradle Runner takes 5 seconds, since Gradle has to do its things. This does not sound a lot if you run it once, but if you are debugging/fixing something this becomes annoying.

Using IntelliJ builder/runner should be much faster and might increase developer joy.

**Implementation ideas**
So I believe that Quarkus Gradle plugin has to be run so everything is setup correctly. But since [Command mode support with IDE launcher](https://github.com/quarkusio/quarkus/pull/7681) is possible, it might be possible to use some similar approach for Unit Tests when run inside IntelliJ.

Also it might be cool if Quarkus tasks would support Gradle build cache, currently they are not cached in 1.3.2.Final. Although that is a different feature.

What do you think, is this possible?

**Comment by asodja** [comment]
Ok. I played around a bit and it seems there are two problems:

1. QuarkusTest expects that compiled classes are in `build/classes/java/main` and `build/classes/java/test` as this is Gradle output folder. Although IntelliJ runner outputs everything to `out` folder. 

**Solution:** I changed project output folder to `build/classes/java/main` and `build/classes/java/test`: 
File -> Project Structure -> Modules -> main module -> Output path:   
`/Users/anzes/prototype-dev/quarkus-gradle/build/classes/java/main`
File -> Project Structure -> Modules -> test module -> Test Output path:   
`/Users/anzes/prototype-dev/quarkus-gradle/build/classes/java/test`.

Quarkus-gradle is my test project. Problem with that is, that whenever you resync Gradle, IntelliJ sets this output back to `/Users/anzes/prototype-dev/quarkus-gradle/out`.

---
2. QuarkusTest tries to reads pom.xml so it can probably run quarkus test setup. I got exception:
```
WARN: Unable to locate the maven project on the filesystem
java.lang.RuntimeException: io.quarkus.bootstrap.BootstrapException: Failed to determine the Maven artifact associated with the application
	at io.quarkus.test.junit.QuarkusTestExtension.beforeEach(QuarkusTestExtension.java:209)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeBeforeEachCallbacks$1(TestMethodTestDescriptor.java:154)

```

**Solution:** I added minimal pom.xml with Quarkus dependencies into project.

---

Now tests are running directly with IntelliJ runner. This is really a hackish solution though and not practical. To make it work you also have to have quarkus dependencies in .m2 folder. 

Nevertheless it seems that there is a way to achieve similar IntelliJ experience for Gradle projects as is for Maven, although 2. step is probably not easy.

**Comment by aloubyansky** [comment]
Yes, we need to do proper discovery using gradle api in this case.

**Comment by jskovjyskebankdk** [comment]
Similar problem in Eclipse, https://github.com/quarkusio/quarkus/issues/9738

I made some work on this a year back that I did not manage to get landed.
I think I will spend some time trying to fix this problem again.

**Comment by aloubyansky** [comment]
FYI, I am finally extracting Maven from the bootstrap-core #9828 and defining a separate quarkus-bootstrap-app-model that both Maven and Gradle bootstrap will implement. Then for Gradle I think we are going to use custom models (e.g. https://github.com/bmuschko/tooling-api-custom-model) to provide the implementation. @glefloch has already volunteered to look into that.

**Comment by asodja** [comment]
One way of doing this would be to just run `Gradle Quarkus plugin` and cache the model. I played a bit with it and made it work for IntelliJ Unit Tests. 

My strategy was smth. like that:
1. Bootstrap checks if classpath has changed (using `System.getProperty("java.class.path").hashCode()`)
2. if classpath changed or model does not exists, then it runs `Gradle quarkus plugin` with some IDE parameters via `Gradle Tooling Api` (I actually added new task to Quarkus plugin, also some workspace discovery is done in this step)
3. Gradle Quarkus plugin then builds the model and replaces paths if needed (e.g. build/classes/java/main with out/production/classes)
4. Bootstrap then reads the new model and saves it as cached model
5. On next run Bootstrap just reuses cached model 

The main benefit is, that Bootstrap reuses Gradle plugin so there is less code in Bootstrap core, but there might be some corner cases I did not think of.

If @glefloch is interested in that approach, working code (last commit from master is from 30 May) can be found here: https://github.com/asodja/quarkus/commit/f0d111d996f4050edf69ade77191e4ae18d72670. Additionally some tests should be added.
